Cryptography API docs (#3319)
dapr/docs: daprdocs/content/en/reference/components-reference/supported-cryptography/local-storage.md

---
type: docs

> title: "Local storage"

Reviewer: Thoughts on adding "(Built-in)" to all the built-in components to make it more clear?

```suggestion
title: "Local storage (Built-in)"
```

Reviewer 2: We have not done this with all the other components. I thinks we only do this once pluggable components gets closer to stable.

Reviewer: In this case, "Built-in" means that it uses the built-in crypto engine in Dapr. So Dapr sees the raw key material. Not sure if there's a better way to explain this, aside from docs

Author: @msfussell @ItalyPaleAle do you think using `builtin.yaml` instead of `generic.yaml` is what's breaking the PR? it worked for me locally 🤷‍♀️

linkTitle: "Local storage"
description: Detailed information on the local storage cryptography component
---

## Component format

Todo: update component format to correct format for cryptography

> 

Reviewer: The purpose of this component is to load keys from a local directory. The component accepts as input the name of a folder, and loads keys from there. Each key is in its own file, and when users request a key with a given name, Dapr will load the file with that name.

A Dapr `crypto.yaml` component file has the following structure:

> 

Reviewer: Here perhaps we can add a note, shared by all built-in crypto components, something like:

```md
This component uses the built-in cryptographic engine in Dapr to perform operations. Although keys are never exposed to your application, Dapr has access to the raw key material.
```

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: mycrypto
spec:
  type: crypto.localstorage
  metadata:
    version: v1
    - name: path
      value: fixtures/crypto/localstorage/
```

{{% alert title="Warning" color="warning" %}}
The above example uses secrets as plain strings. It is recommended to use a secret store for the secrets, as described [here]({{< ref component-secrets.md >}}).
{{% /alert %}}

## Spec metadata fields

| Field              | Required | Details | Example |
|--------------------|:--------:|---------|---------|

> | path               | Y        | Connection-string for the local storage  | `fixtures/crypto/localstorage/`

Reviewer: ```suggestion
| `path`             | Y        | Folder containing the keys to be loaded. When loading a key, the name of the key will be used as name of the file in this folder.  | `/path/to/folder`
```

---
type: docs
title: "Local storage"
linkTitle: "Local storage"
description: Detailed information on the local storage cryptography component
---

## Component format

> Todo: update component format to correct format for cryptography

Reviewer 2: TODO here

A Dapr `crypto.yaml` component file has the following structure:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: mycrypto

>   namespace: default

Reviewer 2: Suggest removing namespace
```suggestion
```

spec:
  type: crypto.localstorage

>   metadata:

Reviewer 2: Missing spec version

```suggestion
  metadata:
       version: v1
```

    - name: path
      value: fixtures/crypto/localstorage/
```

{{% alert title="Warning" color="warning" %}}
The above example uses secrets as plain strings. It is recommended to use a secret store for the secrets, as described [here]({{< ref component-secrets.md >}}).
{{% /alert %}}

## Spec metadata fields

| Field              | Required | Details | Example |
|--------------------|:--------:|---------|---------|

> | path               | Y        | Connection-string for the lcoal storage  | `fixtures/crypto/localstorage/`

Reviewer 2: These are always confusing 
- Is this absolute or relative path or both?
- What is the format of the file? How do I supply the information to decrypt or encrypt?
- Is there a code example anywhere?

Author: @ItalyPaleAle

Reviewer: Looks like it's been solved

---
type: docs
title: "Local storage"
linkTitle: "Local storage"
description: Detailed information on the local storage cryptography component
---

## Component format

> The purpose of this component is to load keys from a local directory. The component accepts as input the name of a folder, and loads keys from there. Each key is in its own file, and when users request a key with a given name, Dapr will load the file with that name.

Reviewer: ```suggestion
The purpose of this component is to load keys from a local directory.

The component accepts as input the name of a folder, and loads keys from there. Each key is in its own file, and when users request a key with a given name, Dapr loads the file with that name.

Supported file formats:

- PEM with public and private keys (supports: PKCS#1, PKCS#8, PKIX)
- JSON Web Key (JWK) containing public, private, or symmetric keys
- Raw key data for symmetric keys
```

{{% alert title="Note" color="primary" %}}
This component uses the cryptographic engine in Dapr to perform operations. Although keys are never exposed to your application, Dapr has access to the raw key material.

{{% /alert %}}

A Dapr `crypto.yaml` component file has the following structure:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: mycrypto
spec:

>   type: crypto.localstorage

Reviewer: ```suggestion
  type: crypto.dapr.localstorage
```

---
type: docs
title: "Local storage"
linkTitle: "Local storage"
description: Detailed information on the local storage cryptography component
---

## Component format

The purpose of this component is to load keys from a local directory.

The component accepts as input the name of a folder, and loads keys from there. Each key is in its own file, and when users request a key with a given name, Dapr loads the file with that name.

Supported file formats:

- PEM with public and private keys (supports: PKCS#1, PKCS#8, PKIX)

> - JSON Web Key (JWK) containing public, private, or symmetric keys

Reviewer: ```suggestion
- JSON Web Key (JWK) containing a public, private, or symmetric key
```

Reviewer: This is because it accepts a single JWK. For more than one key, use the JWKS component.

- Raw key data for symmetric keys

{{% alert title="Note" color="primary" %}}
This component uses the cryptographic engine in Dapr to perform operations. Although keys are never exposed to your application, Dapr has access to the raw key material.

{{% /alert %}}

A Dapr `crypto.yaml` component file has the following structure:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: mycrypto
spec:
  type: crypto.dapr.localstorage
  metadata:
    version: v1
    - name: path

>       value: fixtures/crypto/localstorage/

Reviewer: ```suggestion
      value: /path/to/folder/
```

```

{{% alert title="Warning" color="warning" %}}
The above example uses secrets as plain strings. It is recommended to use a secret store for the secrets, as described [here]({{< ref component-secrets.md >}}).
{{% /alert %}}

## Spec metadata fields

| Field              | Required | Details | Example |
|--------------------|:--------:|---------|---------|

> | `path`             | Y        | Folder containing the keys to be loaded. When loading a key, the name of the key used is the name of the file in this folder.  | `/path/to/folder`

Reviewer 2: You should give an  example here, just to be clear. I say this since you say "keys" plural when I presume this can only be a single key, since the key name is derived from a file. Is this correct, there is only one key per file?

Reviewer: Assume that you set `path=/mnt/keys` and that you have those files in there:

- `/mnt/keys/mykey1.pem`
- `/mnt/keys/mykey2`

When using the component, you can reference keys as `mykey1.pem` and `mykey2`.
